### PR TITLE
specialize io::copy to use copy_file_range, splice or sendfile

### LINKED ISSUE #74426: Specialize std::io::copy for files, sockets, or pipes on Linux
Currently `std::fs::copy` is ~~using `std::io::copy` to copy files - it reads the file content into the memory, and then write it to the output file. `std::fs::copy` or~~ (edit: it tries `copy_file_range(2)` first and has already benefited from the file cloning feature) `std::io::copy` can benefit from the file cloning feature in Btrfs and XFS by using `FICLONE` (and fallbacking to read-write loop if the filesystem does not support it).
I think is easy  to use `FICLONE` in `std::fs::copy`, but it's less general (it requires file paths). `std::io::copy` is more general, but some additional methods may be needed to be added to `std::io::Read` and `std::io::Write` in order to use `FICLONE`.
libuv's support for `FICLONE`: libuv/libuv#1491
The `ioctl_ficlonerange(2)` man page: https://www.man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html

### PR DESCRIPTION
**Pull request by the8472**
Fixes #74426.
Also covers #60689 but only as an optimization instead of an official API.

The specialization only covers std-owned structs so it should avoid the problems with #71091

Currently linux-only but it should be generalizable to other unix systems that have sendfile/sosplice and similar.

There is a bit of optimization potential around the syscall count. Right now it may end up doing more syscalls than the naive copy loop when doing short (<8KiB) copies between file descriptors.

The test case executes the following:

```
[pid 103776] statx(3, "", AT_STATX_SYNC_AS_STAT|AT_EMPTY_PATH, STATX_ALL, {stx_mask=STATX_ALL|STATX_MNT_ID, stx_attributes=0, stx_mode=S_IFREG|0644, stx_size=17, ...}) = 0
[pid 103776] write(4, "wxyz", 4)        = 4
[pid 103776] write(4, "iklmn", 5)       = 5
[pid 103776] copy_file_range(3, NULL, 4, NULL, 5, 0) = 5

```

0-1 `stat` calls to identify the source file type. 0 if the type can be inferred from the struct from which the FD was extracted
𝖬 `write` to drain the `BufReader`/`BufWriter` wrappers. only happen when buffers are present. 𝖬 ≾ number of wrappers present. If there is a write buffer it may absorb the read buffer contents first so only result in a single write. Vectored writes would also be an option but that would require more invasive changes to `BufWriter`.
𝖭 `copy_file_range`/`splice`/`sendfile` until file size, EOF or the byte limit from `Take` is reached. This should generally be *much* more efficient than the read-write loop and also have other benefits such as DMA offload or extent sharing.

## Benchmarks

```

OLD

test io::tests::bench_file_to_file_copy         ... bench:      21,002 ns/iter (+/- 750) = 6240 MB/s    [ext4]
test io::tests::bench_file_to_file_copy         ... bench:      35,704 ns/iter (+/- 1,108) = 3671 MB/s  [btrfs]
test io::tests::bench_file_to_socket_copy       ... bench:      57,002 ns/iter (+/- 4,205) = 2299 MB/s
test io::tests::bench_socket_pipe_socket_copy   ... bench:     142,640 ns/iter (+/- 77,851) = 918 MB/s

NEW

test io::tests::bench_file_to_file_copy         ... bench:      14,745 ns/iter (+/- 519) = 8889 MB/s    [ext4]
test io::tests::bench_file_to_file_copy         ... bench:       6,128 ns/iter (+/- 227) = 21389 MB/s   [btrfs]
test io::tests::bench_file_to_socket_copy       ... bench:      13,767 ns/iter (+/- 3,767) = 9520 MB/s
test io::tests::bench_socket_pipe_socket_copy   ... bench:      26,471 ns/iter (+/- 6,412) = 4951 MB/s
```

### CODE REVIEW

File: library/std/src/io/copy.rs
Comment (the8472) [inline library/std/src/io/copy.rs:404]: Does std have some policy about tests polluting the filesystem? there doesn't seem to be any utility for temporary files and libc doesn't even have a wrapper for `memfd_create` :(

**Comment by KodrAus** [comment]
Thanks for your patience on this so far @the8472! Since this is quite a large PR it'll take me a while to dig through it all.

It's tricky to benchmark IO code, but I would be interested to see roughly what the impact of this is so we can justify the new machinery a bit.

**Comment by the8472** [comment]
@KodrAus it's not just about speed (but that ought to be a benefit too, sure) but also about using `copy_file_range` which reduces storage footprint on copy-on-write filesystems.

But I'll try to come up with a benchmark.

That said, I think #75428, being a bugfix, has a higher priority than this.

**Comment by the8472** [comment]
@KodrAus 

```
ext4
[old] test io::tests::bench_file_to_file_copy         ... bench:      82,997 ns/iter (+/- 7,423) = 1579 MB/s
[new] test io::tests::bench_file_to_file_copy         ... bench:      65,165 ns/iter (+/- 2,448) = 2011 MB/s

btrfs
[old] test io::tests::bench_file_to_file_copy         ... bench:      75,250 ns/iter (+/- 2,551) = 1741 MB/s
[new] test io::tests::bench_file_to_file_copy         ... bench:      10,802 ns/iter (+/- 363) = 12134 MB/s


[old] test io::tests::bench_file_to_pipe_copy         ... bench:      37,602 ns/iter (+/- 19,304) = 3485 MB/s
[new] test io::tests::bench_file_to_pipe_copy         ... bench:      31,220 ns/iter (+/- 15,320) = 4198 MB/s

[old] test io::tests::bench_pipe_to_socket_copy       ... bench:      79,906 ns/iter (+/- 5,964) = 1640 MB/s
[new] test io::tests::bench_pipe_to_socket_copy       ... bench:      54,283 ns/iter (+/- 15,867) = 2414 MB/s
```

All benchmarks copy 128KiB per bencher iteration.  Larger files will probably achieve higher speeds.


While benchmarking I have noticed that results involving pipes can be quite erratic since currently the only form of pipes exposed in the public API is for child process stdio, which means the performance characteristics depend on the process on the other end.
I have used `yes` and `dd` to get the loop of syscalls that I want, but even minor changes can mean one process overtakes another which leads to blocking IO calls suddenly showing drastically different performance. The generic `io::copy` loop uses 1 read and 1 write syscall with an 8k buffer each. While that is in principle slower it can occasionally show higher throughput in benchmarks because `splice` finishes so fast that the source/sink becomes empty/full and the process gets suspended and incurs additional context switch costs while the constant switching between read/write means the process never gets suspended because it is the limiting factor.
`sendfile` from file to pipe works on modern kernels and doesn't seem to suffer from this erratic behavior because it stays in kernel space and just waits when the pipe is full, thus avoiding the context switch overhead.

So I'll go ahead and change the behavior to do syscall bashing to figure out which one works on a pair of file descriptors, with this preference ranking: *copy_file_range > sendfile > splice > read+write*
Modern kernels should then mostly succeed with the first two which provide the most benefits while older kernels will hit the fallbacks more frequently.

**Comment by the8472** [comment]
@KodrAus updated and shaved off some syscalls. You can find benchmarks on top.

**Comment by KodrAus** [comment]
Thanks @the8472! I'll give this a thorough review now :eyes:

**Comment by KodrAus** [comment]
Thanks for your patience on this one @the8472! I just had to confirm that we're ok landing features that depend on specialization again (we had a block on it for a while, but with `min_specialization` things are ok again)

**Comment by the8472** [comment]
Ok, but if you had mentioned that I could have pointed you to other specialization PRs that have landed since. And I am still anticipating the review itself.

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this @the8472!

I've left a few comments to start. I think we should refactor this into the `sys` module, to also avoid having to leak the `CopyResult` types that really only make sense for this specific implementation of `io::copy`.

I haven't dug into the specialized implementations just yet, or the changes to `unix::fs`.

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs:89]: Any suggestions where to put it? Right now it's linux/android-only. With the right `cfg` on various branches it could be extended to other unix systems which have similar syscalls.

Perhaps It could even be extended to windows in so far that the specializations can extract the buffer allocations from `BufReader/Writer` and then pass those buffers to the generic_copy loop for copying larger chunks instead of the default 8k buffer.

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs:451]: This is just a nicer way to call `fstat/statx`. It's equivalent to but more efficient and reliable than `File::open("/proc/self/fd/<fd>")?.metadata()?`

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs:211]: Yes, that's why the enum valuse is called `Fallback`. If that is not obvious it might need a better name.

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs:236]: use fallback == use generic_copy

Comment (the8472) [reply library/std/src/io/copy.rs]: expanded the comment a bit.

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs]: yes

File: library/std/src/io/tests.rs
Comment (the8472) [reply library/std/src/io/tests.rs]: done

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs]: moved it to `sys::unix::kernel_copy` for now.

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs]: no, just an oversight

**Comment by the8472** [comment]
Should be ready for another review pass.

**Review by KodrAus — COMMENTED** [review]
Thanks for working on this! I've done another pass over the code. I think the implementation is good, but I think we should try limit the visibility of the internals needed by the specialized copy as much as possible. Ideally, I think the module structure could look something like this:

```rust
pub mod sys {
    pub mod unix {
        mod copy {
            #[cfg(any(target_os = "linux", target_os = "android"))]
            pub fn copy_fs(from: &Path, to: &Path) -> io::Result<u64> {}

            pub fn copy<R, W>(reader: R, writer: W) -> io::Result<usize> {}
        }

        pub mod fs {
            #[cfg(any(target_os = "linux", target_os = "android"))]
            pub use super::copy::copy_fs as copy;
        }

        pub use self::copy::copy;
    }
}

pub mod io {
    mod copy {
        pub fn copy<R, W>(reader: R, writer: W) -> io::Result<usize> {
            cfg_if::cfg_if! {
                if #[cfg(unix)] {
                    crate::sys::unix::copy(reader, writer)
                } else {
                    copy_fallback(reader, writer)
                }
            }
        }

        pub(crate) fn copy_fallback<R, W>(reader: R, writer: W) -> io::Result<usize> {}
    }

    pub use self::copy::copy;
}
```

That way none of the `CopyResult` or other internal enums need to be visible in the rest of the crate. What do you think?

File: library/std/src/io/copy.rs
Comment (the8472) [reply library/std/src/io/copy.rs:62]: I'm not sure if *fallback* is the right word. From the perspective of the specializations it indeed is the fallback. But the specializations only apply in some narrow scenarios. In the more general case of arbitrary `Read`/`Write` pairs this is the default strategy.

**Comment by the8472** [comment]
@KodrAus 

> I think the module structure could look something like this: [...]

The issue is that there are more implementations of `copy(from: &Path, to: &Path)` in the `unix::fs` module. E.g. a macos one. Should that also be moved into the new `unix::copy` module?

File: library/std/src/lib.rs
Comment (the8472) [reply library/std/src/lib.rs:318]: I only used it for cleanup in tests. But that was before I learned that other IO tests don't clean up the files they create either. So removing it and adding to the clutter instead would also be an option.

**Comment by the8472** [comment]
> That way none of the CopyResult or other internal enums need to be visible in the rest of the crate. What do you think?

I have restricted the visibility of those without reshuffling the other copy methods already present in the `unix::fs` module.

But I'm still lacking a better name for than `generic_copy` or `copy_fallback`. 
Perhaps `userspace_copy`? But that's from the perspective of the specializations that can do kernel offload, not from the perspective of generic `Read` and `Write`.

**Comment by the8472** [comment]
Actually, I can move the linux-only `copy_file_range/sendfile/splice` methods while leaving the multi-platform `copy` wrappers where they are.

**Review by KodrAus — APPROVED** [review]
Nice work @the8472! This looks good to me now 🙂 

Are you happy to squash some of the commit history down a bit? r=me after that!

**Comment by the8472** [comment]
@KodrAus sufficiently squashed?

**Comment by the8472** [comment]
@KodrAus ping

**Comment by KodrAus** [comment]
Thanks @the8472! This has been a long journey, thanks for sticking through it ❤️ This is looking good to me.

@bors r+ rollup=never

**Comment by hyd-dev** [comment]
Android link failure:
```
error: linking with `/android/ndk/arm-14/bin/arm-linux-androideabi-clang` failed: exit code: 1
```
```
std::sys::unix::kernel_copy::sendfile_splice::h9b10b6580108736a: error: undefined reference to 'splice'
```

@the8472 I think my #74426 can be closed after merging this. If you want, you can add "Closes #74426" to the PR description. Thank you!

**Comment by the8472** [comment]
I seems like the android build failure is because the symbol is only exposed for android [API level 21 or higher](https://github.com/aosp-mirror/platform_bionic/commit/065237718a557ac47e9a03521a2d52330199cba8) while CI builds against [API level 14](https://github.com/rust-lang/rust/blob/a601302ff0217b91589b5a7310a8a23adb843fdc/src/ci/docker/host-x86_64/arm-android/Dockerfile#L8)

I'll try the raw syscall then.

**Comment by the8472** [comment]
Added some fixes for android, `./src/ci/docker/run.sh arm-android` passes locally. A side-benefit is that now it won't waste syscalls if seccomp blocks sendfile or splice.

**Comment by KodrAus** [comment]
@the8472 Do you know if we have precedent elsewhere in the standard library for probing syscall availability like this (outside of Windows, where I think it's pretty standard)? It seems slightly suspect to me. If it's already done in a similar way then I'm ok with it, otherwise maybe we should just remove `android` from the`cfg`s.

**Comment by the8472** [comment]
@KodrAus sure, here's a macos example: https://github.com/rust-lang/rust/blob/master/library/std/src/sys/unix/fs.rs#L1327-L1337
and there are a few places where it's done on linux too.

**Comment by KodrAus** [comment]
Thanks @the8472! It looks like it is a pretty standard pattern then. Let’s try run this through again.

@bors r+ rollup=never

**Comment by the8472** [comment]
I can't reproduce that failure locally (running `./src/ci/docker/run.sh dist-x86_64-musl`), even after merging master into the branch. Might be a spurious failure?

**Comment by mati865** [comment]
Quite possible it's spurious: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/segfaults

**Comment by the8472** [comment]
Now it's failing while building LLVM. That very much looks like a problem with the CI runners, as the linked zulip topic says.

**Comment by kennytm** [comment]
@bors retry

https://github.com/rust-lang-ci/rust/runs/1372590716 GCC segfaulted

```
2020-11-09T06:41:38.4040207Z FAILED: sccache /usr/bin/c++   -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Ilib/CodeGen/AsmPrinter -I/checkout/src/llvm-project/llvm/lib/CodeGen/AsmPrinter -Iinclude -I/checkout/src/llvm-project/llvm/include -ffunction-sections -fdata-sections -fPIC -m64 -static-libstdc++ -fPIC -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wno-maybe-uninitialized -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG    -fno-exceptions -fno-rtti -std=c++14 -MMD -MT lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/ARMException.cpp.o -MF lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/ARMException.cpp.o.d -o lib/CodeGen/AsmPrinter/CMakeFiles/LLVMAsmPrinter.dir/ARMException.cpp.o -c /checkout/src/llvm-project/llvm/lib/CodeGen/AsmPrinter/ARMException.cpp
2020-11-09T06:41:38.4045983Z <built-in>: internal compiler error: Segmentation fault
2020-11-09T06:41:38.4046809Z Please submit a full bug report,
2020-11-09T06:41:38.4047510Z with preprocessed source if appropriate.
2020-11-09T06:41:38.4048570Z See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
2020-11-09T06:41:38.9469006Z ninja: build stopped: subcommand failed.
2020-11-09T06:41:38.9573198Z thread 'main' panicked at '
2020-11-09T06:41:38.9573855Z 	finished in 66.701 seconds
2020-11-09T06:41:38.9578641Z command did not execute successfully, got: exit code: 1
2020-11-09T06:41:38.9579350Z 
2020-11-09T06:41:38.9581477Z build script failed, must exit now', /cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.44/src/lib.rs:885:5
```

**Comment by the8472** [comment]
Now it's a network error.

```
warning: spurious network error (2 tries remaining): error inflating zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error inflating zlib stream; class=Zlib (5)
error: failed to get `cc` as a dependency of package `bootstrap v0.0.0 (D:\a\rust\rust\src\bootstrap)`
```

**Comment by the8472** [comment]
Added another commit that incorporates findings from #78979:

* fs::copy now handles EOVERFLOW too instead of asserting it cannot happen. Previously only the io::copy path handled it.
* reduced the chunk size for `copy_file_range` since a user reported that 2GB returns errors on an older kernel but smaller values do work.

**Comment by the8472** [comment]
Thanks @bjorn3, fixed.

**Comment by cuviper** [comment]
Nominating for beta because this missed this 1.49 branch -- otherwise we'll need to repeat the #79008 workaround.

**Comment by Mark-Simulacrum** [comment]
I am dropping the beta nom here, I think I'd rather just land #79008 on beta.

**Pull request merged**
